File: streampark-common/src/main/scala/org/apache/streampark/common/util/HadoopUtils.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.common.util

import org.apache.streampark.common.conf.{CommonConfig, InternalConfigHolder}
import org.apache.streampark.common.conf.ConfigKeys._
import org.apache.streampark.common.util.Implicits._

import org.apache.commons.collections.CollectionUtils
import org.apache.commons.lang3.StringUtils
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs._
import org.apache.hadoop.hdfs.DistributedFileSystem
import org.apache.hadoop.security.UserGroupInformation
import org.apache.hadoop.service.Service.STATE
import org.apache.hadoop.yarn.api.records.{ApplicationId, FinalApplicationStatus, YarnApplicationState}
import org.apache.hadoop.yarn.client.api.YarnClient
import org.apache.hadoop.yarn.conf.YarnConfiguration

import javax.security.auth.kerberos.KerberosTicket

import java.io.{File, IOException}
import java.security.PrivilegedAction
import java.util
import java.util.{Timer, TimerTask}
import java.util.concurrent._

import scala.util.{Failure, Success, Try}

object HadoopUtils extends Logger {

  private[this] lazy val HADOOP_HOME: String = "HADOOP_HOME"
  private[this] lazy val HADOOP_CONF_DIR: String = "HADOOP_CONF_DIR"
  private[this] lazy val CONF_SUFFIX: String = "/etc/hadoop"

  private[this] var reusableYarnClient: YarnClient = _

  private[this] var ugi: UserGroupInformation = _

  private[this] var reusableConf: Configuration = _

  private[this] var reusableHdfs: FileSystem = _

  private[this] var tgt: KerberosTicket = _

  private[this] lazy val configurationCache: util.Map[String, Configuration] =
    new ConcurrentHashMap[String, Configuration]()

  def getUgi(): UserGroupInformation = {
    if (ugi == null) {
      ugi = if (HadoopConfigUtils.kerberosEnable) {
        getKerberosUGI()
      } else {
        UserGroupInformation.createRemoteUser(HadoopConfigUtils.hadoopUserName)
      }
    }
    ugi
  }

  private[this] lazy val hadoopConfDir: String = Try(
    FileUtils.getPathFromEnv(HADOOP_CONF_DIR)) match {
    case Failure(_) => FileUtils.resolvePath(FileUtils.getPathFromEnv(HADOOP_HOME), CONF_SUFFIX)
    case Success(value) => value
  }

  private[this] lazy val tgtRefreshTime: Long = {
    val user = UserGroupInformation.getLoginUser
    val method = classOf[UserGroupInformation].getDeclaredMethod("getTGT")
    method.setAccessible(true)
    tgt = method.invoke(user).asInstanceOf[KerberosTicket]
    Option(tgt) match {
      case Some(value) =>
        val start = value.getStartTime.getTime
        val end = value.getEndTime.getTime
        ((end - start) * 0.90f).toLong
      case _ =>
        logWarn("get kerberos tgtRefreshTime failed, try get kerberos.ttl. ")
        val timeUnit = DateUtils.getTimeUnit(InternalConfigHolder.get(CommonConfig.KERBEROS_TTL))
        timeUnit._2 match {
          case TimeUnit.SECONDS => timeUnit._1 * 1000
          case TimeUnit.MINUTES => timeUnit._1 * 60 * 1000
          case TimeUnit.HOURS => timeUnit._1 * 60 * 60 * 1000
          case TimeUnit.DAYS => timeUnit._1 * 60 * 60 * 24 * 1000
          case _ =>
            throw new IllegalArgumentException(
              s"[StreamPark] parameter:${CommonConfig.KERBEROS_TTL.key} invalided, unit options are [s|m|h|d]")
        }
    }
  }

  def getConfigurationFromHadoopConfDir(confDir: String = hadoopConfDir): Configuration = {
    if (!configurationCache.containsKey(confDir)) {
      FileUtils.exists(confDir)
      val hadoopConfDir = new File(confDir)
      val confName = List("hdfs-default.xml", "core-site.xml", "hdfs-site.xml", "yarn-site.xml")
      val files =
        hadoopConfDir
          .listFiles()
          .filter(x => x.isFile && confName.contains(x.getName))
          .toList
      val conf = new HadoopConfiguration()
      if (CollectionUtils.isNotEmpty(files)) {
        files.foreach(x => conf.addResource(new Path(x.getAbsolutePath)))
      }
      configurationCache.put(confDir, conf)
    }
    configurationCache(confDir)
  }

  /**
   * <pre> Note: There are two ways to load a hadoop configuration file:<br> 1) Copy
   * core-site.xml,hdfs-site.xml,yarn-site.xml of hadoop conf to the resource dir.<br> 2)
   * Automatically goes to $HADOOP_HOME/etc/hadoop to load the configuration.<br> We recommend the
   * second method, without copying the configuration files.<br> </pre>
   */
  def hadoopConf: Configuration = Option(reusableConf).getOrElse {
    reusableConf = getConfigurationFromHadoopConfDir(hadoopConfDir)
    // add hadoopConfDir to classpath...you know why???
    ClassLoaderUtils.loadResource(hadoopConfDir)

    if (StringUtils.isBlank(reusableConf.get("hadoop.tmp.dir"))) {
      reusableConf.set("hadoop.tmp.dir", "/tmp")
    }
    if (StringUtils.isBlank(reusableConf.get("hbase.fs.tmp.dir"))) {
      reusableConf.set("hbase.fs.tmp.dir", "/tmp")
    }
    // disable timeline service as we only query yarn app here.
    // Otherwise we may hit this kind of ERROR:
    // java.lang.ClassNotFoundException: com.sun.jersey.api.client.config.ClientConfig
    reusableConf.set("yarn.timeline-service.enabled", "false")
    reusableConf.set("fs.hdfs.impl", classOf[DistributedFileSystem].getName)
    reusableConf.set("fs.file.impl", classOf[LocalFileSystem].getName)
    reusableConf.set("fs.hdfs.impl.disable.cache", "true")
    reusableConf
  }

  private[this] def closeHadoop(): Unit = {
    if (reusableHdfs != null) {
      reusableHdfs.close()
      reusableHdfs = null
    }
    if (reusableYarnClient != null) {
      reusableYarnClient.close()
      reusableYarnClient = null
    }
    if (tgt != null && !tgt.isDestroyed) {
      tgt.destroy()
      tgt = null
    }
    reusableConf = null
    ugi = null
  }

  private[this] def getKerberosUGI(): UserGroupInformation = {
    logInfo("kerberos login starting....")

    require(
      HadoopConfigUtils.kerberosPrincipal.nonEmpty && HadoopConfigUtils.kerberosKeytab.nonEmpty,
      s"$KEY_SECURITY_KERBEROS_PRINCIPAL and $KEY_SECURITY_KERBEROS_KEYTAB must not be empty")

    System.setProperty("javax.security.auth.useSubjectCredsOnly", "false")

    if (HadoopConfigUtils.kerberosKrb5.nonEmpty) {
      System.setProperty("java.security.krb5.conf", HadoopConfigUtils.kerberosKrb5)
      System.setProperty("java.security.krb5.conf.path", HadoopConfigUtils.kerberosKrb5)
    }

    System.setProperty("sun.security.spnego.debug", HadoopConfigUtils.kerberosDebug)
    System.setProperty("sun.security.krb5.debug", HadoopConfigUtils.kerberosDebug)
    hadoopConf.set(KEY_HADOOP_SECURITY_AUTHENTICATION, KEY_KERBEROS)

    Try {
      UserGroupInformation.setConfiguration(hadoopConf)
      val ugi =
        UserGroupInformation.loginUserFromKeytabAndReturnUGI(
          HadoopConfigUtils.kerberosPrincipal,
          HadoopConfigUtils.kerberosKeytab)
      UserGroupInformation.setLoginUser(ugi)
      logInfo("kerberos authentication successful")
      ugi
    } match {
      case Success(ugi) => ugi
      case Failure(e) => throw e
    }
  }

  def hdfs: FileSystem = {
    Option(reusableHdfs).getOrElse {
      reusableHdfs = Try {
        getUgi().doAs[FileSystem](new PrivilegedAction[FileSystem]() {
          // scalastyle:off FileSystemGet
          override def run(): FileSystem = FileSystem.get(hadoopConf)
          // scalastyle:on FileSystemGet
        })
      } match {
        case Success(fs) =>
          if (HadoopConfigUtils.kerberosEnable) {
            // reLogin...
            val timer = new Timer()
            timer.schedule(
              new TimerTask {
                override def run(): Unit = {
                  closeHadoop()
                  logInfo(
                    s"Check Kerberos Tgt And reLogin From Keytab Finish:refresh time: ${DateUtils.format()}")
                }
              },
              tgtRefreshTime,
              tgtRefreshTime)
          }
          fs
        case Failure(e) =>
          throw new IllegalArgumentException(s"[StreamPark] access hdfs error: $e")
      }
      reusableHdfs
    }
  }

  def yarnClient: YarnClient = {
    if (reusableYarnClient == null || !reusableYarnClient.isInState(STATE.STARTED)) {
      reusableYarnClient = Try {
        getUgi().doAs(new PrivilegedAction[YarnClient]() {
          override def run(): YarnClient = {
            val yarnConf = new YarnConfiguration(hadoopConf);
            val client = YarnClient.createYarnClient;
            client.init(yarnConf);
            client.start();
            client
          }
        })
      } match {
        case Success(client) => client
        case Failure(e) =>
          throw new IllegalArgumentException(s"[StreamPark] access yarnClient error: $e")
      }
    }
    reusableYarnClient
  }

  def toApplicationId(appId: String): ApplicationId = {
    require(
      appId != null,
      "[StreamPark] HadoopUtils.toApplicationId: applicationId muse not be null")
    val timestampAndId = appId.split("_")
    ApplicationId.newInstance(timestampAndId(1).toLong, timestampAndId.last.toInt)
  }

  @throws[IOException]
  def downloadJar(jarOnHdfs: String): String = {
    val tmpDir = FileUtils.createTempDir()
    val fs = FileSystem.get(new Configuration)
    val sourcePath = fs.makeQualified(new Path(jarOnHdfs))
    if (!fs.exists(sourcePath)) {
      throw new IOException(s"jar file: $jarOnHdfs doesn't exist.")
    }
    val destPath = new Path(tmpDir.getAbsolutePath + "/" + sourcePath.getName)
    fs.copyToLocalFile(sourcePath, destPath)
    new File(destPath.toString).getAbsolutePath
  }

  def toYarnState(state: String): YarnApplicationState = {
    YarnApplicationState.values.find(_.name() == state).orNull
  }

  def toYarnFinalStatus(state: String): FinalApplicationStatus = {
    FinalApplicationStatus.values.find(_.name() == state).orNull
  }

  private class HadoopConfiguration extends Configuration {

    private lazy val rewriteNames = List(
      "dfs.blockreport.initialDelay",
      "dfs.datanode.directoryscan.interval",
      "dfs.heartbeat.interval",
      "dfs.namenode.decommission.interval",
      "dfs.namenode.replication.interval",
      "dfs.namenode.checkpoint.period",
      "dfs.namenode.checkpoint.check.period",
      "dfs.client.datanode-restart.timeout",
      "dfs.ha.log-roll.period",
      "dfs.ha.tail-edits.period",
      "dfs.datanode.bp-ready.timeout")

    private def getHexDigits(value: String): String = {
      var negative = false
      var str = value
      var hexString: String = null
      if (value.startsWith("-")) {
        negative = true
        str = value.substring(1)
      }
      if (str.startsWith("0x") || str.startsWith("0X")) {
        hexString = str.substring(2)
        if (negative) hexString = "-" + hexString
        return hexString
      }
      null
    }

    // HDFS default value change (with adding time unit) breaks old version MR tarball work with Hadoop 3.x
    // detail: https://issues.apache.org/jira/browse/HDFS-12920
    private def getSafeValue(name: String): String = {
      val value = getTrimmed(name)
      if (rewriteNames.contains(name)) {
        return value.replaceFirst("s$", "")
      }
      value
    }

    override def getLong(name: String, defaultValue: Long): Long = {
      val valueString = getSafeValue(name)
      valueString match {
        case null => defaultValue
        case v =>
          val hexString = getHexDigits(v)
          if (hexString != null) {
            return java.lang.Long.parseLong(hexString, 16)
          }
          java.lang.Long.parseLong(valueString)
      }
    }
  }
}

File: streampark-common/src/main/java/org/apache/streampark/common/util/HadoopUtils.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.common.util;

import org.apache.streampark.common.conf.CommonConfig;
import org.apache.streampark.common.conf.ConfigKeys;
import org.apache.streampark.common.conf.InternalConfigHolder;

import org.apache.streampark.shaded.org.slf4j.Logger;

import org.apache.commons.collections.CollectionUtils;
import org.apache.commons.lang3.StringUtils;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.LocalFileSystem;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.hdfs.DistributedFileSystem;
import org.apache.hadoop.security.UserGroupInformation;
import org.apache.hadoop.service.Service;
import org.apache.hadoop.yarn.api.records.ApplicationId;
import org.apache.hadoop.yarn.api.records.FinalApplicationStatus;
import org.apache.hadoop.yarn.api.records.YarnApplicationState;
import org.apache.hadoop.yarn.client.api.YarnClient;
import org.apache.hadoop.yarn.conf.YarnConfiguration;

import javax.security.auth.kerberos.KerberosTicket;

import java.io.File;
import java.io.IOException;
import java.lang.reflect.Method;
import java.security.PrivilegedAction;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.ConcurrentHashMap;

/** Hadoop filesystem and YARN client utilities. */
public final class HadoopUtils {

    private static final Logger LOG =
        StreamParkLoggerFactory.loggerFactory().getLogger(HadoopUtils.class.getName());

    private static final String HADOOP_HOME = "HADOOP_HOME";
    private static final String HADOOP_CONF_DIR = "HADOOP_CONF_DIR";
    private static final String CONF_SUFFIX = "/etc/hadoop";

    private static YarnClient reusableYarnClient;
    private static UserGroupInformation ugi;
    private static Configuration reusableConf;
    private static FileSystem reusableHdfs;
    private static KerberosTicket tgt;

    private static final Map<String, Configuration> CONFIGURATION_CACHE = new ConcurrentHashMap<>();

    private static String hadoopConfDir;
    private static Long tgtRefreshTime;

    private HadoopUtils() {
    }

    public static UserGroupInformation getUgi() {
        if (ugi == null) {
            if (HadoopConfigUtils.KERBEROS_ENABLE) {
                ugi = getKerberosUGI();
            } else {
                ugi = UserGroupInformation.createRemoteUser(HadoopConfigUtils.HADOOP_USER_NAME);
            }
        }
        return ugi;
    }

    private static String getHadoopConfDir() {
        if (hadoopConfDir == null) {
            try {
                hadoopConfDir = FileUtils.getPathFromEnv(HADOOP_CONF_DIR);
            } catch (Exception e) {
                try {
                    hadoopConfDir =
                        FileUtils.resolvePath(FileUtils.getPathFromEnv(HADOOP_HOME), CONF_SUFFIX);
                } catch (Exception ignored) {
                    hadoopConfDir = "";
                }
            }
        }
        return hadoopConfDir;
    }

    private static long getTgtRefreshTime() {
        if (tgtRefreshTime == null) {
            try {
                UserGroupInformation user = UserGroupInformation.getLoginUser();
                Method method = UserGroupInformation.class.getDeclaredMethod("getTGT");
                method.setAccessible(true);
                tgt = (KerberosTicket) method.invoke(user);
                if (tgt != null) {
                    long start = tgt.getStartTime().getTime();
                    long end = tgt.getEndTime().getTime();
                    tgtRefreshTime = (long) ((end - start) * 0.90f);
                } else {
                    LOG.warn("[StreamPark] get kerberos tgtRefreshTime failed, try get kerberos.ttl.");
                    DateUtils.TimeUnitPair timeUnit =
                        DateUtils.getTimeUnit(InternalConfigHolder.get(CommonConfig.KERBEROS_TTL()));
                    switch (timeUnit.unit) {
                        case SECONDS:
                            tgtRefreshTime = (long) timeUnit.num * 1000;
                            break;
                        case MINUTES:
                            tgtRefreshTime = (long) timeUnit.num * 60 * 1000;
                            break;
                        case HOURS:
                            tgtRefreshTime = (long) timeUnit.num * 60 * 60 * 1000;
                            break;
                        case DAYS:
                            tgtRefreshTime = (long) timeUnit.num * 60 * 60 * 24 * 1000;
                            break;
                        default:
                            throw new IllegalArgumentException(
                                "[StreamPark] parameter:"
                                    + CommonConfig.KERBEROS_TTL().getKey()
                                    + " invalided, unit options are [s|m|h|d]");
                    }
                }
            } catch (Exception e) {
                throw new IllegalStateException("Failed to resolve kerberos TTL configuration", e);
            }
        }
        return tgtRefreshTime;
    }

    public static Configuration getConfigurationFromHadoopConfDir() {
        return getConfigurationFromHadoopConfDir(getHadoopConfDir());
    }

    public static Configuration getConfigurationFromHadoopConfDir(String confDir) {
        return CONFIGURATION_CACHE.computeIfAbsent(
            confDir,
            dir -> {
                FileUtils.exists(dir);
                File hadoopConfDirFile = new File(dir);
                List<String> confName =
                    Arrays.asList(
                        "hdfs-default.xml",
                        "core-site.xml",
                        "hdfs-site.xml",
                        "yarn-site.xml");
                File[] allFiles = hadoopConfDirFile.listFiles();
                HadoopConfiguration conf = new HadoopConfiguration();
                if (allFiles != null && CollectionUtils.isNotEmpty(Arrays.asList(allFiles))) {
                    for (File file : allFiles) {
                        if (file.isFile() && confName.contains(file.getName())) {
                            conf.addResource(new Path(file.getAbsolutePath()));
                        }
                    }
                }
                return conf;
            });
    }

    public static Configuration hadoopConf() {
        if (reusableConf == null) {
            reusableConf = getConfigurationFromHadoopConfDir(getHadoopConfDir());
            try {
                ClassLoaderUtils.loadResource(getHadoopConfDir());
            } catch (Exception e) {
                throw new IllegalStateException("Failed to load Hadoop resources from " + getHadoopConfDir(), e);
            }

            if (StringUtils.isBlank(reusableConf.get("hadoop.tmp.dir"))) {
                reusableConf.set("hadoop.tmp.dir", "/tmp");
            }
            if (StringUtils.isBlank(reusableConf.get("hbase.fs.tmp.dir"))) {
                reusableConf.set("hbase.fs.tmp.dir", "/tmp");
            }
            reusableConf.set("yarn.timeline-service.enabled", "false");
            reusableConf.set("fs.hdfs.impl", DistributedFileSystem.class.getName());
            reusableConf.set("fs.file.impl", LocalFileSystem.class.getName());
            reusableConf.set("fs.hdfs.impl.disable.cache", "true");
        }
        return reusableConf;
    }

    private static void closeHadoop() {
        if (reusableHdfs != null) {
            try {
                reusableHdfs.close();
            } catch (IOException e) {
                LOG.warn("[StreamPark] close hdfs failed", e);
            }
            reusableHdfs = null;
        }
        if (reusableYarnClient != null) {
            try {
                reusableYarnClient.close();
            } catch (IOException e) {
                LOG.warn("[StreamPark] close yarn client failed", e);
            }
            reusableYarnClient = null;
        }
        if (tgt != null && !tgt.isDestroyed()) {
            try {
                tgt.destroy();
            } catch (javax.security.auth.DestroyFailedException e) {
                LOG.warn("[StreamPark] destroy kerberos ticket failed", e);
            }
            tgt = null;
        }
        reusableConf = null;
        ugi = null;
    }

    private static UserGroupInformation getKerberosUGI() {
        LOG.info("[StreamPark] kerberos login starting....");

        if (HadoopConfigUtils.KERBEROS_PRINCIPAL.isEmpty()
            || HadoopConfigUtils.KERBEROS_KEYTAB.isEmpty()) {
            throw new IllegalArgumentException(
                ConfigKeys.KEY_SECURITY_KERBEROS_PRINCIPAL()
                    + " and "
                    + ConfigKeys.KEY_SECURITY_KERBEROS_KEYTAB()
                    + " must not be empty");
        }

        System.setProperty("javax.security.auth.useSubjectCredsOnly", "false");

        if (!HadoopConfigUtils.KERBEROS_KRB5.isEmpty()) {
            System.setProperty("java.security.krb5.conf", HadoopConfigUtils.KERBEROS_KRB5);
            System.setProperty("java.security.krb5.conf.path", HadoopConfigUtils.KERBEROS_KRB5);
        }

        System.setProperty("sun.security.spnego.debug", HadoopConfigUtils.KERBEROS_DEBUG);
        System.setProperty("sun.security.krb5.debug", HadoopConfigUtils.KERBEROS_DEBUG);
        hadoopConf()
            .set(
                ConfigKeys.KEY_HADOOP_SECURITY_AUTHENTICATION(),
                ConfigKeys.KEY_KERBEROS());

        try {
            UserGroupInformation.setConfiguration(hadoopConf());
            UserGroupInformation kerberosUgi =
                UserGroupInformation.loginUserFromKeytabAndReturnUGI(
                    HadoopConfigUtils.KERBEROS_PRINCIPAL, HadoopConfigUtils.KERBEROS_KEYTAB);
            UserGroupInformation.setLoginUser(kerberosUgi);
            LOG.info("[StreamPark] kerberos authentication successful");
            return kerberosUgi;
        } catch (IOException e) {
            throw new IllegalStateException("Kerberos authentication failed", e);
        }
    }

    public static FileSystem hdfs() {
        if (reusableHdfs == null) {
            try {
                reusableHdfs =
                    getUgi()
                        .doAs(
                            (PrivilegedAction<FileSystem>) () -> {
                                try {
                                    return FileSystem.get(hadoopConf());
                                } catch (IOException e) {
                                    throw new IllegalStateException("Failed to obtain HDFS FileSystem", e);
                                }
                            });
                if (HadoopConfigUtils.KERBEROS_ENABLE) {
                    Timer timer = new Timer();
                    long refreshTime = getTgtRefreshTime();
                    timer.schedule(
                        new TimerTask() {

                            @Override
                            public void run() {
                                closeHadoop();
                                LOG.info(
                                    "[StreamPark] Check Kerberos Tgt And reLogin From Keytab Finish:refresh time: {}",
                                    DateUtils.format());
                            }
                        },
                        refreshTime,
                        refreshTime);
                }
            } catch (Exception e) {
                throw new IllegalArgumentException("[StreamPark] access hdfs error: " + e, e);
            }
        }
        return reusableHdfs;
    }

    public static YarnClient yarnClient() {
        if (reusableYarnClient == null || !reusableYarnClient.isInState(Service.STATE.STARTED)) {
            try {
                reusableYarnClient =
                    getUgi()
                        .doAs(
                            (PrivilegedAction<YarnClient>) () -> {
                                YarnConfiguration yarnConf =
                                    new YarnConfiguration(hadoopConf());
                                YarnClient client = YarnClient.createYarnClient();
                                client.init(yarnConf);
                                client.start();
                                return client;
                            });
            } catch (Exception e) {
                throw new IllegalArgumentException(
                    "[StreamPark] access yarnClient error: " + e, e);
            }
        }
        return reusableYarnClient;
    }

    public static ApplicationId toApplicationId(String appId) {
        if (appId == null) {
            throw new IllegalArgumentException(
                "[StreamPark] HadoopUtils.toApplicationId: applicationId muse not be null");
        }
        String[] timestampAndId = appId.split("_");
        return ApplicationId.newInstance(
            Long.parseLong(timestampAndId[1]), Integer.parseInt(timestampAndId[timestampAndId.length - 1]));
    }

    public static String downloadJar(String jarOnHdfs) throws IOException {
        File tmpDir = FileUtils.createTempDir();
        FileSystem fs = FileSystem.get(new Configuration());
        Path sourcePath = fs.makeQualified(new Path(jarOnHdfs));
        if (!fs.exists(sourcePath)) {
            throw new IOException("jar file: " + jarOnHdfs + " doesn't exist.");
        }
        Path destPath = new Path(tmpDir.getAbsolutePath() + "/" + sourcePath.getName());
        fs.copyToLocalFile(sourcePath, destPath);
        return new File(destPath.toString()).getAbsolutePath();
    }

    public static YarnApplicationState toYarnState(String state) {
        for (YarnApplicationState yarnState : YarnApplicationState.values()) {
            if (yarnState.name().equals(state)) {
                return yarnState;
            }
        }
        return null;
    }

    public static FinalApplicationStatus toYarnFinalStatus(String state) {
        for (FinalApplicationStatus status : FinalApplicationStatus.values()) {
            if (status.name().equals(state)) {
                return status;
            }
        }
        return null;
    }

    private static final class HadoopConfiguration extends Configuration {

        private static final List<String> REWRITE_NAMES =
            Arrays.asList(
                "dfs.blockreport.initialDelay",
                "dfs.datanode.directoryscan.interval",
                "dfs.heartbeat.interval",
                "dfs.namenode.decommission.interval",
                "dfs.namenode.replication.interval",
                "dfs.namenode.checkpoint.period",
                "dfs.namenode.checkpoint.check.period",
                "dfs.client.datanode-restart.timeout",
                "dfs.ha.log-roll.period",
                "dfs.ha.tail-edits.period",
                "dfs.datanode.bp-ready.timeout");

        private static String getHexDigits(String value) {
            boolean negative = false;
            String str = value;
            if (value.startsWith("-")) {
                negative = true;
                str = value.substring(1);
            }
            if (str.startsWith("0x") || str.startsWith("0X")) {
                String hexString = str.substring(2);
                if (negative) {
                    hexString = "-" + hexString;
                }
                return hexString;
            }
            return null;
        }

        private String getSafeValue(String name) {
            String value = getTrimmed(name);
            if (REWRITE_NAMES.contains(name)) {
                return value.replaceFirst("s$", "");
            }
            return value;
        }

        @Override
        public long getLong(String name, long defaultValue) {
            String valueString = getSafeValue(name);
            if (valueString == null) {
                return defaultValue;
            }
            String hexString = getHexDigits(valueString);
            if (hexString != null) {
                return Long.parseLong(hexString, 16);
            }
            return Long.parseLong(valueString);
        }
    }
}
